### PR TITLE
MULE-19226: Policy not able to inject internal dependencies for artifacts that shares with the application

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 1]
```diff
@@ -464,9 +465,9 @@ public static ExtensionModel loadExtension(Class<?> clazz, Map<String, Object> p
     return new DefaultJavaExtensionModelLoader().loadExtensionModel(clazz.getClassLoader(), dslResolvingContext, params);
   }
 
-  public static String getImplicitConfigurationProviderName(ExtensionModel extensionModel,
+  public static String getImplicitConfigurationProviderName(MuleContext muleContext, ExtensionModel extensionModel,
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:468]: get the id (better called `prefix`) directly instead of passing the `MuleContext` around

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java [hunk 2]
```diff
@@ -58,8 +58,8 @@ public CompositeArtifactExtensionManager(ExtensionManager parentExtensionManager
     this.childExtensionManager = childExtensionManager;
 
     extensionModels = new HashSet<>();
-    extensionModels.addAll(parentExtensionManager.getExtensions());
     extensionModels.addAll(childExtensionManager.getExtensions());
+    extensionModels.addAll(parentExtensionManager.getExtensions());
```
Comment (marianogonzalez) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java:62]: why this change? this looks extra suspicious considering you're adding into a HashSet

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java [hunk 3]
```diff
@@ -132,12 +132,10 @@ public ConfigurationInstance getConfiguration(String configurationProviderName,
   public Optional<ConfigurationProvider> getConfigurationProvider(ExtensionModel extensionModel, ComponentModel componentModel) {
     Optional<ConfigurationProvider> configurationModel =
         childExtensionManager.getConfigurationProvider(extensionModel, componentModel);
-
-    if (!configurationModel.isPresent()) {
+    if (!configurationModel.isPresent() && !childExtensionManager.getExtension(extensionModel.getName()).isPresent()) {
```
Comment (marianogonzalez) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java:135]: pretty sure this breaks backwards compatibility when a config is defined at the domain level

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/policy/PolicyTemplate.java [hunk 4]
```diff
@@ -26,13 +26,8 @@
    */
   void dispose();
 
-  /**
-   * @return plugins deployed only inside the policy template
-   */
-  List<ArtifactPlugin> getArtifactPlugins();
```
Comment (marianogonzalez) [inline modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/policy/PolicyTemplate.java:32]: Can't remove this. This is API

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 5]
```diff
@@ -464,9 +465,10 @@ public static ExtensionModel loadExtension(Class<?> clazz, Map<String, Object> p
     return new DefaultJavaExtensionModelLoader().loadExtensionModel(clazz.getClassLoader(), dslResolvingContext, params);
   }
 
-  public static String getImplicitConfigurationProviderName(ExtensionModel extensionModel,
+  public static String getImplicitConfigurationProviderName(String prefix, ExtensionModel extensionModel,
                                                             ConfigurationModel implicitConfigurationModel) {
-    return format("%s-%s-implicit", extensionModel.getName(), implicitConfigurationModel.getName());
+    Preconditions.checkArgument(isNotEmpty(prefix), "A prefix must be provided");
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:470]: static import

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 6]
```diff
@@ -464,9 +465,10 @@ public static ExtensionModel loadExtension(Class<?> clazz, Map<String, Object> p
...
+    return format("%s-%s-%s-implicit", prefix, extensionModel.getName(), implicitConfigurationModel.getName());
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:471]: doesn't this break for the domain use case?

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java (hunk 2 above)
Comment (IvanAndresFritzler) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java:62]: Classloaders for each extension are set using a property of their extension models. This change aims to use the policy classloader when the policy is executing and not the application one (note that the set uses the extension name as the key, so the order matters, because in the case of an application declaring the same extension than a policy, the policy extension model was not added to this set).

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/policy/PolicyTemplate.java (hunk 4 above)
Comment (IvanAndresFritzler) [reply modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/policy/PolicyTemplate.java:32]: It's not removed (see that I also renamed getOwnArtifactPlugins to getArtifactPlugins)

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java (hunk 2 above)
Comment (marianogonzalez) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java:62]: This is not the correct way of achieving what you want:

* This is a HashSet, order is not guaranteed. This is would actually be flaky on different depending on the JDK version/vendor to use
* If you want a specific classloader to be used, then you need to reference it specifically (usually by setting it as the TCCL).

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java (hunk 3 above)
Comment (marianogonzalez) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java:135]: You can have the case in which both the child and parent both define an extension, but the child uses it implicitly while the parent uses it explicitly. In that case, one of two things can happen:

1) This change is superflous as it ends up looking into the parent anyway
2) it returns an empty Optional when before it would had returned the parents provider, therefore breaking backwards compatibility

**Comment by IvanAndresFritzler** [comment]
--validate

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java (hunk 2 above)
Comment (IvanAndresFritzler) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java]: After a refactor in order to address the backward compatibility issue, this discussion no longer applies.

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java (hunk 3 above)
Comment (IvanAndresFritzler) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/CompositeArtifactExtensionManager.java]: The code has been refactored and the CompositeArtifactExtensionManager is no longer modified.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java (hunk 6 above)
Comment (IvanAndresFritzler) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java]: Yes, it was breaking the domain use case. Code has been refactored in order to avoid this.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java [hunk 7]
```diff
@@ -134,6 +142,18 @@ protected void validateOperationConfiguration(ConfigurationProvider configuratio
     }
   }
 
+  private void validatePolicyIsolation(ConfigurationProvider configurationProvider) {
+    if (featureFlaggingService.isEnabled(MuleRuntimeFeature.ENABLE_POLICY_ISOLATION)
+        && muleContext.getArtifactType().equals(ArtifactType.POLICY)
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java:147]: Static imports

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/MuleArtifactContext.java [hunk 8]
```diff
@@ -803,7 +803,14 @@ private static void configureSplitterExceptionHandlingFeature() {
     FeatureFlaggingRegistry ffRegistry = FeatureFlaggingRegistry.getInstance();
 
     ffRegistry.registerFeature(HANDLE_SPLITTER_EXCEPTION,
-                               ctx -> ctx.getConfiguration().getMinMuleVersion().map(v -> v.atLeast("4.4.0")).orElse(false));
+                               muleContext -> {
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/MuleArtifactContext.java:806]: why?

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/model/properties/PropertiesResolverUtils.java [hunk 9]
```diff
@@ -341,7 +341,14 @@ public static void configurePropertiesResolverFeatureFlag() {
       FeatureFlaggingRegistry ffRegistry = FeatureFlaggingRegistry.getInstance();
 
       ffRegistry.registerFeature(HONOUR_RESERVED_PROPERTIES,
-                                 ctx -> ctx.getConfiguration().getMinMuleVersion().map(v -> v.newerThan("4.2.2")).orElse(false));
+                                 muleContext -> {
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/model/properties/PropertiesResolverUtils.java:344]: why?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ImplicitConfigurationProviderFactory.java [hunk 10]
```diff
@@ -41,4 +41,17 @@ ConfigurationProvider createImplicitConfigurationProvider(ExtensionModel extensi
...
+  /**
+   * Returns an implicit {@link ConfigurationProvider} name, that can be used to look it up.
+   *
+   * @param extensionModel     The configurable {@link ExtensionModel}.
+   * @param configurationModel The {@link ConfigurationModel} that represents the extensionModel configuration.
+   * @param muleContext        The corresponding {@link MuleContext}.
+   * @return The {@link ConfigurationProvider} name.
+   * @see org.mule.runtime.module.extension.internal.manager.ExtensionRegistry#getConfigurationProvider(String)
+   */
+  String resolveImplicitConfigurationProviderName(ExtensionModel extensionModel, ConfigurationModel configurationModel,
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ImplicitConfigurationProviderFactory.java:54]: add `@since`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java [hunk 11]
```diff
@@ -113,4 +119,47 @@ private boolean needsDynamicConnectionProvider(ExtensionModel extensionModel, Co
...
+   * {@link MuleRuntimeFeature#ENABLE_POLICY_ISOLATION} feature flag.
+   * 
+   * @param extensionModel     The configurable {@link ExtensionModel}.
+   * @param configurationModel The {@link ConfigurationModel} that represents the extensionModel configuration.
+   * @param muleContext        The corresponding {@link MuleContext}.
+   * @return The {@link ConfigurationProvider} name.
+   */
+  private String featureFlaggedPolicyExtensionsImplicitConfigurationProviderName(ExtensionModel extensionModel,
+                                                                                 ConfigurationModel configurationModel,
+                                                                                 MuleContext muleContext) {
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java:147]: have the mule context injected into this class instead of receiving it as parameter

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManager.java [hunk 12]
```diff
@@ -86,6 +86,11 @@
   public void initialise() throws InitialisationException {
     extensionRegistry = new ExtensionRegistry(new DefaultRegistry(muleContext));
     extensionActivator = new ExtensionActivator(muleContext);
+    try {
+      initialiseIfNeeded(implicitConfigurationProviderFactory, muleContext);
+    } catch (MuleException e) {
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManager.java:91]: catch the more specific `InitialisationException`

Comment (elrodro83) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManager.java:91]: and actaully, no need to catch since this method is already throwing that!

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java (hunk 11 above)
Comment (elrodro83) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java:147]: review all the chained calls where the mule context is being passed around

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java [hunk 13]
```diff
@@ -134,6 +142,18 @@ protected void validateOperationConfiguration(ConfigurationProvider configuratio
     }
   }
 
+  private void validatePolicyIsolation(ConfigurationProvider configurationProvider) {
+    if (featureFlaggingService.isEnabled(ENABLE_POLICY_ISOLATION)
+        && muleContext.getArtifactType().equals(POLICY)
+        && configurationProvider.getExtensionModel() != extensionModel) {
+      throw new IllegalOperationException(format(
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java:149]: add a TODO to make this validation at the AST level, before actually initializing the app: https://www.mulesoft.org/jira/browse/MULE-19350

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by IvanAndresFritzler** [comment]
--validate

**Review by elrodro83 — DISMISSED** [review]

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ImplicitConfigurationProviderFactory.java [hunk 14]
```diff
@@ -39,6 +37,17 @@ ConfigurationProvider createImplicitConfigurationProvider(ExtensionModel extensi
...
+  /**
+   * Returns an implicit {@link ConfigurationProvider} name, that can be used to look it up.
+   *
+   * @param extensionModel     The configurable {@link ExtensionModel}.
+   * @param configurationModel The {@link ConfigurationModel} that represents the extensionModel configuration.
+   * @return The {@link ConfigurationProvider} name.
+   * @see org.mule.runtime.module.extension.internal.manager.ExtensionRegistry#getConfigurationProvider(String)
+   * @since 4.4.0
+   */
+  String resolveImplicitConfigurationProviderName(ExtensionModel extensionModel, ConfigurationModel configurationModel);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ImplicitConfigurationProviderFactory.java:51]: why is this moved into this factory and exposed in this interface? it's weird that we have the need to expose this as an interface method. Sounds like an implementation detail that shouldn't be exposed by a factory class. This is a code smell. Double check if this responsibility doesn't belong somewhere else. I don't understand the reason why it should be moved here

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java [hunk 15]
```diff
@@ -137,6 +136,65 @@ private void initPolicyContext() throws InitialisationException {
...
+
+  /**
+   * HTTP and Sockets extension models must be added if found in the parent artifact extensions (backward compatibility for API
+   * Gateway).
+   * 
+   * @return Set containing the parent artifact HTTP and Sockets extension models (if present).
+   */
+  private Set<ExtensionModel> getInheritedExtensionModels() {
+    Set<ExtensionModel> inheritedExtensionModels = new HashSet<>(2);
+    application.getRegistry().<ExtensionManager>lookupByName(OBJECT_EXTENSION_MANAGER).get().getExtension("HTTP")
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java:191]: avoid looking up the extension manager twice in the registry

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManager.java [hunk 16]
```diff
@@ -76,6 +78,9 @@
   @Inject
   private ExpressionManager expressionManager;
 
+  @Inject
+  FeatureFlaggingService featureFlaggingService;
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManager.java:82]: private

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java [hunk 17]
```diff
@@ -41,58 +43,63 @@
  */
 public final class DefaultImplicitConfigurationProviderFactory implements ImplicitConfigurationProviderFactory {
 
+  @Inject
+  MuleContext muleContext;
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java:47]: private

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java [hunk 18]
```diff
@@ -41,58 +43,63 @@
...
+  @Inject
+  FeatureFlaggingService featureFlaggingService;
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConfigurationProviderFactory.java:49]: private

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 19]
```diff
@@ -125,6 +130,9 @@
  */
 public class MuleExtensionUtils {
 
+  private static final String IMPLICIT_CONFIGURATION_SUFFIX = "-implicit";
+  public static final String HYPHEN = "-";
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:134]: isnt there a definition for this somewhere else already?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 20]
```diff
@@ -125,6 +130,9 @@
  */
 public class MuleExtensionUtils {
 
+  private static final String IMPLICIT_CONFIGURATION_SUFFIX = "-implicit";
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:133]: `HYPHEN + "implicit"`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 21]
```diff
@@ -481,9 +488,68 @@ public static ExtensionModel loadExtension(Class<?> clazz, Map<String, Object> p
...
+   *
+   * @param extensionModel     The configurable {@link ExtensionModel}.
+   * @param configurationModel The {@link ConfigurationModel} that represents the extension configuration.
+   * @return The {@link org.mule.runtime.extension.api.runtime.config.ConfigurationProvider} name.
+   */
+  private static String getFeatureFlaggedPolicyExtensionsImplicitConfigurationProviderName(ExtensionModel extensionModel,
+                                                                                           ConfigurationModel configurationModel,
+                                                                                           String extendedArtifactId,
+                                                                                           FeatureFlaggingService featureFlaggingService) {
+    if (featureFlaggingService != null && featureFlaggingService.isEnabled(ENABLE_POLICY_ISOLATION)) {
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java:525]: can `featureFlaggingService ` be null?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManagerTestCase.java [hunk 22]
```diff
@@ -151,7 +152,7 @@
   @Before
   public void before() throws MuleException {
     muleContext = mockContextWithServices();
-
+    when(muleContext.getArtifactType()).thenReturn(ArtifactType.APP);
```
Comment (elrodro83) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/manager/DefaultExtensionManagerTestCase.java:155]: static import

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java [hunk 23]
```diff
@@ -137,6 +136,66 @@ private void initPolicyContext() throws InitialisationException {
     }
   }
 
+  /**
+   * Applies the {@link MuleRuntimeFeature#ENABLE_POLICY_ISOLATION} feature to the policy artifact plugins list.
+   *
+   * @return The policy artifact plugins.
+   */
+  private List<ArtifactPlugin> getFeatureFlaggedArtifactPlugins(ArtifactDescriptor policyArtifactDescriptor) {
+    if (FeatureFlaggingUtils.isFeatureEnabled(ENABLE_POLICY_ISOLATION, policyArtifactDescriptor)) {
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java:145]: static import

**Review by elrodro83 — APPROVED** [review]

**Review by marianogonzalez — DISMISSED** [review]

**Review by marianogonzalez — APPROVED** [review]

**Review by elrodro83 — APPROVED** [review]
